src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+
+It doesn't look like much, but, when porting libraries to a new platform, getting to the stage where it produces a valid output for the first time is pretty exciting!
+
+## Using the port
+
+As mentioned [above](#heading=h.mogcyiy1x8w8), the port depends on a few Emscripten features that currently need to be enabled at the linking stage of the application. If you want to use this libusb port in your own application, here's what you'll need to do:
+
+1. Download libusb fork from [https://github.com/RReverser/libusb](https://github.com/RReverser/libusb) either as an archive as part of your build or add it as a git submodule in your project.
+2. Run `autoreconf -fiv` in the `libusb` folder.
+3. Run `emconfigure ./configure –host=wasm32 –prefix=/some/installation/path` to initialize the project for cross-compilation and to set a path where you want to put the built artifacts.
+4. Run `emmake make install`.
+5. Point your application or higher-level library to search for the libusb under the earlier chosen path.
+6. Add the following flags to your application's link arguments: `--bind -s ASYNCIFY -s ALLOW_MEMORY_GROWTH`.
+
+The library currently has a few limitations:
+
+1. No transfer cancellation support. This is a limitation of WebUSB, which, in turn, stems from lack of cross-platform transfer cancellation in libusb itself.
+2. No isochronous transfer support. It shouldn't be hard to add it by following the implementation of existing transfer modes as examples, but it's also a somewhat rare mode and I didn't have any devices to test it on, so for now I left it as unsupported. If you do have such devices, and want to contribute to the library, PRs are welcome!
+3. The [earlier mentioned cross-platform limitations](#heading=h.ik9rifeitesl). Those limitations are imposed by operating systems, so not much we can do here, except ask users to override the driver or permissions. However, if you're porting HID or serial devices, you can follow the libusb example and port some other library to another Fugu API. For example, you could port a C library [hidapi](https://github.com/libusb/hidapi) to [WebHID](/hid/) and side-step those issues, associated with low-level USB access, altogether.
+
+## Conclusion
+
+In this post I've shown how, with the help of Emscripten, Asyncify and Fugu APIs even low-level libraries like libusb can be ported to the web with a few integration tricks.
+
+Porting such essential and widely used low-level libraries is particularly rewarding, because, in turn, it allows to bring higher-level libraries or even whole applications to the web, too. This opens experiences that were previously limited to users of one or two platforms, to all kinds of devices and operating systems, making those experiences available just a link click away.
```

Comment:
```suggestion
Porting such essential and widely used low-level libraries is particularly rewarding, because, in turn, it allows to bringing higher-level libraries or even whole applications to the web, too. This opens experiences that were previously limited to users of one or two platforms, to all kinds of devices and operating systems, making those experiences just a link click away.
```

Reply:
Hmm "it allows to bringing" doesn't sound right grammatically.

Reply:
Yeah, I messed that suggestion up. I removed 'to'.

Reply:
Why did you remove "available" btw? Without it that sentence doesn't sound right either.

Reply:
It doesn't sound weird to me. 'Just a click away' means available, making the word redundant.

Reply:
Right but together with the verb before it reads as "making them a click away" which, to me, sounds like it's possible to _make_ them easily rather than _access_ them easily, so "available" or "accessible" feels necessary.

Reply:
'Making' has [two definitions in English](https://www.google.com/search?q=making+definition&oq=making+definition&aqs=chrome..69i57j0i512l2j69i60j69i65l3j69i61.2539j1j7&sourceid=chrome&ie=UTF-8). This is entirely idiomatic.

Reply:
Ah. Perhaps it's just one of those things that sound fine to native speakers who are aware of the secondary meaning, but that non-native speakers like me might stumble upon when reading. I can remove the word if you insist though.